OffHeap Changes

rmnattas opened the pull request:
Changes to support OffHeap

* Add pinning array support for OffHeap by storing the pinning array in node extension.
* Add OffHeap support in constrainAnyIntLoad for valuePropagation
* Update LoopVersioner code to support OffHeap

rmnattas commented:
Failure in CI JitBuilder testing, looking into it
```
Assertion failed at /home/vsts/work/1/s/compiler/il/OMRNode.cpp:212: self()->hasSymbolReference() + self()->hasRegLoadStoreSymbolReference() + self()->hasBranchDestinationNode() + self()->hasBlock() + self()->hasArrayStride() + self()->hasPinningArrayPointer() + self()->hasDataType() <= 1
_unionPropertyA union is not disjoint for this node aloadi (0x55a674c83880):
has({SymbolReference, ...}, ..., DataType) = ({1,0},0,0,0,1,0)
```

rmnattas commented:
Fixed

vijaysun-omr commented on compiler/il/OMRNode.cpp:
```diff
@@ -5046,10 +5046,16 @@ OMR::Node::hasArrayStride()
    return self()->getOpCode().isArrayLength();
    }
 
+bool
```
Can you please clarify why the support for pinning array in node extension is only enabled for data address pointer ? i.e. why not use it for all kinds of nodes ?

VermaSh replied:
The limitation was sort of picked up while trying to mimic `hasPinningArrayPointer()`behavior*. But we can remove the limitation by checking extension element size instead:
```diff
- OMR::Node::supportsPinningArrayPointerInNodeExtension()
+ OMR::Node::hasPinningArrayPointerInNodeExtension()
   {
- return self()->isDataAddrPointer();
+   int extensionElemNum = _unionBase._extension.getNumElems();
+   return (extensionElemNum >= 6 && self().getPinningArrayPointer() != NULL)
   }
```

\*`hasPinningArrayPointer()`, unlike the name suggests, it doesn't actually check if the node has a pinning array pointer, It just checks if the opCode supports it. A way to replicate that for node extensions was by checking for some sort of flag on the node, which in this case happened to be `isDataAddrPointer()`.

rmnattas replied:
Updated code to support pinning array in node extension for all nodes as suggested 👍

vijaysun-omr replied:
Thanks, please run as much internal testing as you are able to, to ensure that that change does not cause failures in downstream projects.

rmnattas replied:
Change to support pinning array for node extension for all nodes results in failures that needs investigation, opened an issue https://github.com/eclipse/omr/issues/7379 to track that work to be done later.

vijaysun-omr commented on compiler/optimizer/VPHandlers.cpp:
```diff
@@ -1376,7 +1376,15 @@ TR::Node *constrainAnyIntLoad(OMR::ValuePropagation *vp, TR::Node *node)
                uintptr_t offset = vp->comp()->target().is64Bit() ? (uintptr_t)index->getUnsignedLongInt() : (uintptr_t)index->getUnsignedInt();
-               uintptr_t chIdx = (offset - (uintptr_t)TR::Compiler->om.contiguousArrayHeaderSizeInBytes()) / 2;
+               uintptr_t chIdx = 0;
+               if (array->isDataAddrPointer())
```
It may be better to explain the reasoning here via some comments (including the original "divide by 2" logic).

rmnattas replied:
Added comment "Get index from offset of a 2-byte element array"

vijaysun-omr replied:
Thanks, but I meant also why the below math is different for data address pointer and why it works out in the "if" and "else" cases.

vijaysun-omr commented:
For the loop versioner commit, I would like @jdmpapin to also review.

There is a typo in that loop versioner commit message : "to use the chilid of aloadi tree"

vijaysun-omr commented on compiler/optimizer/VPHandlers.cpp:
```diff
@@ -2019,77 +2028,114 @@ static TR::Node *findArrayIndexNode(OMR::ValuePropagation *vp, TR::Node *node, i
+         offset2 = offset;
+         }
+      else
+         offset2 = offset->getFirstChild();
```
More comments about the shape of the trees in these two cases, the 'if' and 'else' here would be good. For example, why is it okay to proceed with `offset2 == offset` in the `if` case ? More importantly, why is it okay to call `getFirstChild` on `offset` here ? i.e. do we know that `offset` has children ?

VermaSh replied:
Here's some context for that logic:
Expected tree structure
1. Some constant needs to be added to the offset
```
aladd
    aloadi (dataAddrPointer)
        aload (array object)
    ladd/lsub
        lmul
            i2l
                iload array_index
            lconst mul_stride/shift_amount
    lconst some_constant
```
2. No constant needs to be added to the offset
```
aladd
    aloadi (dataAddrPointer)
        aload (array object)
    lmul
        i2l
            iload array_index
        lconst mul_stride
```
3. Stride == 1
```
aladd
    aloadi (dataAddrPointer)
        aload (array object)
    iload array_index
```
`Offset2` is just an alias for `offset` node which points to `lmul/lshl` or `array_index` node.

We can proceed with `offset2 == offset` because we assume that either it is array_index (case 3) or it has array_index as a child of `lmul` node.

> why is it okay to call getFirstChild on offset on the else side ? i.e. do we know that offset has children ?

We expect offset to have a child based on the expected tree structures. Either the child is `lmul` or `array_index` node.

We should add a check for these assumptions. @rmnattas While we are at it can you add a assert failure for all other node types? I added a check for shift node but it doesn't look like it is handled correctly here.

rmnattas replied:
Thank you @VermaSh, not seeing what case would result in 1. (some constant being added beside offset) as non-offheap case doesn't cover that?

rmnattas replied:
Added comments to explain the trees in the code.

rmnattas commented:
I updated the VPHandler code and added comments, please review changes.
Also added a `i2l` check for the `stride==1` case which didn't exist originally but I believe is needed.

vijaysun-omr commented:
The additional comments and changes help. The code looks somewhat easier to follow. I'll wait for Devin's review as well

jdmpapin commented on compiler/optimizer/LoopVersioner.cpp:
```diff
@@ -8023,6 +8023,9 @@ bool TR_LoopVersioner::requiresPrivatization(TR::Node *node)
       return false;
 
+   if (node->isDataAddrPointer())
+      return false;
```
Most privatizations will be due to `requiresPrivatization()`, i.e. because they appear to be necessary, and this change will prevent them :+1:

However, I'm concerned that versioner might still [heuristically decide to hoist](https://github.com/eclipse/omr/blob/5d2a72e72daeada0b2d205aafb7df5f082a16151/compiler/optimizer/LoopVersioner.cpp#L5737-L5761) a data address load

Could I ask you to try a small OpenJ9 test case and check the log? [DoesItHoist.java](https://github.com/user-attachments/files/15929423/DoesItHoist.java.txt)
```
$ export TR_EnableExpensiveOptsAtWarm=1
$ java -Xjit:'limit=go,count=63,disableAsyncCompilation,initialOptLevel=warm,inhibitRecompilation,disableSuffixLogs,compilationThreads=1,traceFull,log=go.log' DoesItHoist
```
This should privatize the entire load `inst.next.next.next.next.x[32]`, but *not* any internal pointer (`aladd`, data address) needed to perform that load. When I run it without off-heap arrays, it avoids privatizing the `aladd`, as can be seen in the log:
```
n311n     istore  <temp slot 10>[#441  Auto]        <---- privatizing the element load
n310n       iloadi  <array-shadow>[#229  Shadow]
n309n         aladd (internalPtr )                  <---- aladd correctly not privatized
n307n           aload  <temp slot 9>[#440  Auto]    <---- privatized array reference
n308n           lconst 136
```
I think what prevents the privatization here is that `aladd` doesn't satisfy `isSupportedForPRE()`. But `aloadi` does, so maybe the data address will still get privatized. If it does, it looks like it should be possible to prevent that by handling data address pointers in [`addressPointsAtObject()`](https://github.com/eclipse/omr/blob/5d2a72e72daeada0b2d205aafb7df5f082a16151/compiler/il/OMRNode.cpp#L2500-L2515)

I tried to run this with off-heap myself, but I couldn't figure out how to enable it in my build

rmnattas replied:
It seem to work correctly, I tested it with a build that has all offheap changes (including https://github.com/eclipse/omr/pull/7368 which has code to disable privatization of the data address too).

```
n365n     BBStart <block_42> (freq 158)                                                       [     0x3fcc59861d0] bci=[-1,20,22] rc=0 vc=0 vn=- li=- udi=- nc=0
n340n     istore  <temp slot 10>[#417  Auto] [flags 0x3 0x0 ]                                 [     0x3fcc5985a00] bci=[-1,46,23] rc=0 vc=0 vn=- li=- udi=- nc=1
n339n       iloadi  <array-shadow>[#254  Shadow] [flags 0x80000603 0x0 ]                      [     0x3fcc59859b0] bci=[-1,46,23] rc=1 vc=0 vn=- li=- udi=- nc=1
n338n         aladd (internalPtr )                                                            [     0x3fcc5985960] bci=[-1,46,23] rc=1 vc=0 vn=- li=- udi=- nc=2 flg=0x8000
n336n           aloadi  <contiguousArrayDataAddrFieldSymbol>[#355  final Shadow +8] [flags 0x20604 0x0 ] (internalPtr )  [     0x3fcc59858c0] bci=[-1,46,23] rc=1 vc=0 vn=- li=- udi=- nc=1 >
n335n             aload  <temp slot 9>[#416  Auto] [flags 0x7 0x0 ]                           [     0x3fcc5985870] bci=[-1,41,23] rc=1 vc=0 vn=- li=- udi=- nc=0
n337n           lconst 128                                                                    [     0x3fcc5985910] bci=[-1,46,23] rc=1 vc=0 vn=- li=- udi=- nc=0
n366n     BBEnd </block_42> =====                                                             [     0x3fcc5986220] bci=[-1,20,22] rc=0 vc=0 vn=- li=- udi=- nc=0
```

jdmpapin replied:
OK, thanks, I see the change in `isSupportedNodeForFunctionality()` that makes this work

jdmpapin replied:
I'm OK with it, knowing that's coming in the other PR

jdmpapin reviewed (commented):
Loop versioner changes LGTM, but there may be a small piece missing. I also have a couple of comments about `findArrayIndexNode()`

rmnattas commented on compiler/optimizer/VPHandlers.cpp:
```diff
@@ -2019,77 +2028,133 @@ static TR::Node *findArrayIndexNode(OMR::ValuePropagation *vp, TR::Node *node, i
-     int32_t constValue;
-     if ((offset->getOpCode().isAdd() || offset->getOpCode().isSub()) &&
+   // Note: index node here refers to the array element index.
+   if (node->getFirstChild()->isDataAddrPointer())
```
I noticed that but given that the code was already done it made sense to merge it to make the helper correct for OffHeap if used.

jdmpapin replied:
In that case, would you mind deleting this function instead?

rmnattas replied:
Make sense, removed keeping a comment for possibly constraint code in the future.
https://github.com/eclipse/omr/commit/bb0b3e0d3b531a1dd0157adbf851f426f960b125

jdmpapin replied:
Thanks!

rmnattas commented on compiler/optimizer/VPHandlers.cpp:
```diff
@@ -2019,77 +2028,133 @@ static TR::Node *findArrayIndexNode(OMR::ValuePropagation *vp, TR::Node *node, i
+         //         iload array_index
+         //       lconst mul_stride/shift_amount
+         //     lconst some_constant
+         offset2 = offset->getFirstChild();
```
I noticed that it ignores the addition, given it matches the `usingAladd`/non-offheap path it made sense to keep it consistent.

jdmpapin replied:
It isn't consistent though. The off-heap path ignores any outer addition or subtraction, which can cause it to return a node that doesn't represent the right index. The existing code OTOH ignores an outer addition or subtraction only if it effectively adds exactly the array header size, which means that it will never miss a constant term in the index. The corresponding thing for the off-heap case would be to ignore an outer addition or subtraction of exactly zero, since the data address pointer already points to the first element

Regardless, given https://github.com/eclipse/omr/pull/7345#discussion_r1649334281, IMO this should be deleted rather than correctled anyway

vijaysun-omr commented:
Jenkins build all

vijaysun-omr commented:
Checks have passed. Merging.
